src/Numerics/LinearSolvers/BatchedGeneralizedMinimalResidualSolver.jl

```diff
@@ -0,0 +1,824 @@
+@inline convert_structure!(x, y::MPIStateArray, reshape_tuple, permute_tuple) =
+    convert_structure!(x, y.data, reshape_tuple, permute_tuple)
+@inline convert_structure!(x::MPIStateArray, y, reshape_tuple, permute_tuple) =
+    convert_structure!(x.data, y, reshape_tuple, permute_tuple)
+
+# Kernels
+"""
+initialize_gmres_kernel!(gmres)
+
+# Description
+Initializes the gmres struct by calling
+- initialize_arnoldi
+- initialize_QR!
+- update_arnoldi!
+- update_QR!
+- solve_optimization!
+It is assumed that the first two krylov vectors are already constructed
+
+# Arguments
+- `gmres`: (struct) gmres struct
+
+# Return
+(implicitly) kernel abstractions function closure
+"""
+@kernel function initialize_gmres_kernel!(gmres)
+    I = @index(Global)
+    initialize_arnoldi!(gmres, I)
+    update_arnoldi!(1, gmres, I)
+    initialize_QR!(gmres, I)
+    update_QR!(1, gmres, I)
+    solve_optimization!(1, gmres, I)
+end
+
+"""
+gmres_update_kernel!(i, gmres, I)
+
+# Description
+kernel that calls
+- update_arnoldi!
+- update_QR!
+- solve_optimization!
+Which is the heart of the gmres algorithm
+
+# Arguments
+- `i`: (int) interation index
+- `gmres`: (struct) gmres struct
+- `I`: (int) thread index
+
+# Return
+kernel object from KernelAbstractions
+"""
+@kernel function gmres_update_kernel!(i, gmres)
+    I = @index(Global)
+    update_arnoldi!(i, gmres, I)
+    update_QR!(i, gmres, I)
+    solve_optimization!(i, gmres, I)
+end
+
+"""
+construct_solution_kernel!(i, gmres)
+
+# Description
+given step i of the gmres iteration, constructs the "best" solution of the linear system for the given Krylov subspace
+
+# Arguments
+- `i`: (int) gmres iteration
+- `gmres`: (struct) gmres struct
+
+# Return
+kernel object from KernelAbstractions
+"""
+@kernel function construct_solution_kernel!(i, gmres)
+    M, I = @index(Global, NTuple)
+    tmp = zero(eltype(gmres.b))
+    @inbounds for j in 1:i
+        tmp += gmres.Q[M, j, I] * gmres.sol[j, I]
+    end
+    gmres.x[M, I] += tmp # since previously gmres.x held the initial value
+end
+
+# Configuration for Kernels
+"""
+initialize_gmres!(gmres; ndrange = gmres.n, cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Uses the initialize_gmres_kernel! for initalizing
+
+# Arguments
+- `gmres`: (struct) [OVERWRITTEN]
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function initialize_gmres!(
+    gmres::BatchedGeneralizedMinimalResidual;
+    ndrange = gmres.n,
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = initialize_gmres_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = initialize_gmres_kernel!(CUDA(), gpu_threads)
+    end
+    event = kernel!(gmres, ndrange = ndrange)
```

Comment:
I think you might want to synchronize with the default stream here by passing an `Event(CUDA())` or `Event(CPU())` as a `dependencies` key word argument.